### ISSUE TITLE
How to run PostgreSQL on VPS with High Availability?

### ISSUE THREAD

**Issue by nelsonic** [dwyl/learn-devops#58]
Linode has a Postgres App in the "Marketplace":
https://www.linode.com/marketplace/apps/linode/postgresql
It runs on Debian which is _fine_ to start with.
I would definitely prefer to run it on OpenBSD: https://github.com/dwyl/learn-security/issues/73

# Requirements

+ [ ] High Security
+ [ ] Low maintenance. 
  + [ ] Full documentation for setup and maintenance
+ [ ] At least two nodes (master+replica in separate data centres) so we can have failover.
+ [ ] Lower total cost than AWS RDS https://aws.amazon.com/rds/postgresql/pricing

**Comment by nelsonic** [comment]
I'm bumping this up to **`priority-1`** because this is now _costing_ us/me money each month and I _feel_ it. #62 💸 🔥 
Going to spend a couple of hours on this _today_.

**Comment by nelsonic** [comment]
At the time of writing, Linode (_the hosting service we are using for Hits_) does not have a _managed_ database product.
Digital ocean has such a product: https://www.digitalocean.com/pricing/#managed-databases

![digital-ocean-managed-databases](https://user-images.githubusercontent.com/194400/87876439-00155500-c9d0-11ea-8e93-9985927a228c.png)


But the _minimum_ we would pay is **$50**/month given how much data (**`25GB`**) we _already_ have in our Hits app. 
That would be borderline "OK" in the short term, but the second we need slightly more disk space, 
the next level up is **`38`** GB and the price _doubles_ to **$100**/month spending $1200/year on a "hobby" app is an *eye-watering* amount of money!!

As for the AWS RDS pricing you practically need a PhD to understand it! https://aws.amazon.com/rds/aurora/pricing
![image](https://user-images.githubusercontent.com/194400/87876665-a0b84480-c9d1-11ea-962d-41c68f1f5a2d.png)

The _cheapest_ instance that can be used AWS RDS Postgres is `db.t3.medium` which costs **`$0.093`/hour**
This is **`$67.89`**/month for _each_ instance. So the _minimum_ we would be paying (_with replication_) would be **`$135.78`**/month which is daylight robbery. 

And don't even get me _started_ on the "Serverless" option: https://aws.amazon.com/rds/aurora/serverless
That's the definition of a ripoff! 
+ $0.06 per ACU-hour
+ $0.10 per GiB-month
Imagine you use 1 "ACU" for the full month (_duh!_) and 25GB of storage, it's: $43.8 + $2.50 = **$46.30**/month
That's not even unreasonable on the surface. But for _basic_ apps that require the database to be "always on",
 it's silly that the _minimum_ charge is **$44/month**. We can do *much* better by hand-crafting a DB setup on VPS.

**Comment by nelsonic** [comment]
Thinking of using the Block Storage: https://www.linode.com/products/block-storage
![image](https://user-images.githubusercontent.com/194400/87877073-23420380-c9d4-11ea-9164-b89bb8f112dc.png)

The storage is independent of the instance(s) so we can use a 1GB ($5/month) Linode instance and unlimited storage!
![image](https://user-images.githubusercontent.com/194400/87877180-ba0ec000-c9d4-11ea-8013-5f597ec07349.png)

**3x Replication**:
![image](https://user-images.githubusercontent.com/194400/87877234-0bb74a80-c9d5-11ea-8f36-88791bad4277.png)

"_You can also increase the size of a block storage volume at any time. This gives you the flexibility to update your storage capacity as your needs change with a few easy steps_."

This is _exactly_ what I want for a DB!

According to this thread: https://www.linode.com/community/questions/18893/how-do-i-move-my-database-files-to-a-block-storage-volume it's possible to use Block Storage as the data directory in Postgres. 

It's going to require a couple of hours of DevOps (_documenting the setup_) but I think it will be _worth_ it!
We can easily run the Hits App for **`$10/month` with **`50GB`** of Disk space. And the only additional cost per month will be the disk allocation.

**Comment by nelsonic** [comment]
AWS has a Amazon Elastic Block Store which we have used in the past, it was one of their earliest products. But it does _not_ have replication or high availability by default. You need to _pay_ extra for that: https://aws.amazon.com/ebs/pricing
![aws-ebs-pricing](https://user-images.githubusercontent.com/194400/87877596-238fce00-c9d7-11ea-9411-3c0b8747fcf4.png)

**`$0.116`**/GB-month for the EBS and **`$0.053`**/GB-month for the EBS Snapshots (_backups_) = **`0.169`**/GB-month i.e. 69% more expensive per GB than Linode. Doesn't seem like a good idea.

Then we would still need to use an AWS EC2 instance where the price for a comparable instance (**`1GB RAM`**) being a **`t3.micro`** is **`$0.0118`**/hour or **`8.614`**/month  https://aws.amazon.com/ec2/pricing/on-demand 
![Screenshot 2020-07-19 at 15 50 54](https://user-images.githubusercontent.com/194400/87877701-afa1f580-c9d7-11ea-9eb1-4598733b954d.png)

We could _attempt_ to run it on a **`t3.nano`** with **`500mb RAM`** at  **`$0.0059`**/hour  or **`4.307`**/month, but I have my doubts as to the performance of such an instance. 😕 
Read this: https://www.cloudsqueeze.ai/amazons-t3-who-should-use-it-when-how-and-the-why/index.html
![image](https://user-images.githubusercontent.com/194400/87877869-b9782880-c9d8-11ea-8d7c-347ac95e5686.png)

Basically you will get an "unexpected bill for additional virtual CPU" when you go over the measly 2 VCPU baseline and it's charged at **`0.05`**/hour i.e. more than the cost of the instance!!! So imagine people are online for **16h/day** and they are making requests that tax our DB server's CPU, the additional CPU would be 16h x $0.05 x 365 / 12 = **`$24/month`**
They charge you for the **_full_ hour** where the CPU spike occurred, so even if there are higher requests in 1 _minute_ per hour we will pay for the **_full_ hour**. I think this rules out AWS four our "Hits" App purposes. 

Going to proceed with my exploration of Linode Block storage. 🔍

**Comment by nelsonic** [comment]
On further reading, it turns out that DigitalOcean Block storage is based on SSDs and they have at-rest encryption:
https://www.digitalocean.com/products/block-storage

![digitalocean-block-storage-ssds](https://user-images.githubusercontent.com/194400/87880366-11b72680-c9e9-11ea-9198-3eaced8a29b3.png)

And they have a _tutorial_ for migrating the Postgres data directory to their Block storage:
https://www.digitalocean.com/community/tutorials/how-to-move-a-postgresql-data-directory-to-a-new-location-on-ubuntu-18-04 (_not just a forum topic, an actual step-by-step tutorial_).

I think the performance and reliability benefits of SSD are worth it.
https://www.enterprisestorageforum.com/storage-hardware/ssd-vs-hdd.html
![image](https://user-images.githubusercontent.com/194400/87880431-84280680-c9e9-11ea-83e1-5e3d07d94bce.png)

Going to revive my DigitialOcean account and try to run a Postgres instance with Block Storage.

**Comment by nelsonic** [comment]
Briefly considered using this Supabase Postgres marketplace image:
https://cloud.digitalocean.com/marketplace/5eb4c8436e9007280dbcd8a3?i=933568
![image](https://user-images.githubusercontent.com/194400/87888273-cec67480-ca23-11ea-811c-81ae38a6e08b.png)

But while Supabase looks _sick_! see: https://github.com/supabase/supabase
As previously mentioned in: https://github.com/dwyl/how-to-choose-a-database/issues/14
It's still considered "Alpha" ... 
![image](https://user-images.githubusercontent.com/194400/87888542-6a0c1980-ca25-11ea-873a-6cb91e4393e4.png)

So I think I'm just going to do a manual / first principals setup of Postgres on Ubuntu 18.04 and document it.

**Comment by nelsonic** [comment]
No other Postgres available in the DigitalOcean Marketplace: https://marketplace.digitalocean.com

![digital-ocean-supabase-postgres](https://user-images.githubusercontent.com/194400/87888715-57deab00-ca26-11ea-9df5-bbb63951c3ba.png)

Going to install Postgres and do writeup. 📝

**Comment by nelsonic** [comment]
Select the instance Operating System and Plan:
![do-create-01-select-instance-type](https://user-images.githubusercontent.com/194400/87891845-4a7bed80-ca33-11ea-8bf5-eb3095eeafe6.png)

Scroll down to the "Add Block Storage" section and enter 1gb into the field:
![do-create-02-block-storage](https://user-images.githubusercontent.com/194400/87891890-667f8f00-ca33-11ea-9043-cca1b44cc334.png)

We're going to increase the size in a later step, so keep it low for now just to set it up.
https://www.digitalocean.com/docs/volumes/how-to/increase-size/

Select `Ext4` for the filesystem and then select the datacenter region that bests suits your needs.

**Comment by nelsonic** [comment]
Scroll down until you see the "Choose a hostname" field. Enter a relevant name, in our case "hits":
![do-create-04-hostname](https://user-images.githubusercontent.com/194400/87892697-b3646500-ca35-11ea-854a-364a4ae0a964.png)

Scroll to the bottom of the page and click "**Create Droplet**".

**Comment by nelsonic** [comment]
## Progress update: 
This morning I created a new DO droplet (VPS instance) with attached block storage.
Installed Postgres on the instance and changed the data directory of Postgres to be the block storage. 
This is all documented in: https://github.com/dwyl/hits/pull/82
This is technically _not_ "High Availability" because there is only _one_ Postgres instance.
If the Postgres data directory were to be compromised or corrupted, we would lose _everything_.
But for the purposes of "hits" we can already do the `filesystem` to `postgres` migration with this setup.

> **Note**: I will still _definitely_ pursue the HA route to have at least _two_ Postgres nodes, one as a "hot" backup.
Obviously this _doubles_ the cost of running the service. But it's the only way to have a backup. Or is it ...?
Considering using DO Object Storage: #65 

My _next_ step is to install NGINX on the droplet and open port 80/443. 
Luckily we've done this before: https://github.com/dwyl/learn-microsoft-azure#12-install--run-nginx-to-test

**Comment by nelsonic** [comment]
Continue: https://cloud.digitalocean.com/droplets/200671195/volumes?i=933568

**Comment by nelsonic** [comment]
Fly.io have made this a _lot_ simpler: https://fly.io/docs/reference/postgres/
We have already used it to deploy the Hits App https://hits.dwyl.com/ see: https://github.com/dwyl/hits/pull/128
Lowering the priority of this issue to **`priority-4`** ("_maybe someday_") and removing it from the App backlog.

**Comment by nelsonic** [comment]
Closing as for now we are using Fly.io 🚀 
Will re-open if/when needed. 👌
